File: binary-search/binary-search.php
<?php
// Define keys to illustrate the correct key is being returned.
$numbers = [
    0 => 1,
    1 => 2,
    2 => 3,
    3 => 4,
    4 => 5,
    5 => 6,
    6 => 7,
    8 => 9,
    9 => 10,
    10 => 11,
];

/**
 * Finds the needle in the haystack and returns the key.
 *
 * @param array $haystack
 * @param int $needle
 * @param int $startPoint
 * @param int $endPoint
 * @return int
 */
function binarySearch(array $haystack, $needle, $startPoint = 0, $endPoint = -1)
{
    if ($endPoint === -1) {
        $endPoint = sizeof($haystack) -1;
    }

    $diff =  $endPoint - $startPoint;
    $jump = (int) floor($diff / 2);
    $searchKey = $startPoint +  $jump;
    $item = $haystack[$searchKey];

    if ($needle === $item) {
        return $searchKey;
    } elseif ($item > $needle) {
      return binarySearch($haystack, $needle, $startPoint, $searchKey -1);
    } else {
      return binarySearch($haystack, $needle, $searchKey + 1, $endPoint);
    }
}

var_dump(binarySearch($numbers, 4));

File: algorithms/binary-search/binary-search.go
package main

import (
  "fmt"
  "math"
)

func binary_search(haystack []int, needle int) int {
  return binary_search_int(haystack, needle, 0, len(haystack))
}

func binary_search_int(haystack []int, needle int, startPoint int, endPoint int) int {

    diff := endPoint - startPoint
    jump :=  int(math.Floor(float64(diff / 2)))
    searchKey := startPoint + jump

    item := haystack[searchKey]
    if (item == needle) {
      return searchKey
    } else if (item > needle) {
      return binary_search_int(haystack, needle, startPoint, searchKey - 1)
    }

    return binary_search_int(haystack, needle, searchKey + 1, endPoint)

}

func main() {
    numbers := []int{1,2,3,4,5,6,7,8,9,10}
    needle := 4
    searchKey := binary_search(numbers, needle)
    fmt.Println(fmt.Sprintf("Number %d found at %d", needle, searchKey))

}
